core/json/src/main/scala/net/liftweb/json/JsonParser.scala

```diff
@@ -187,35 +226,38 @@ object JsonParser {
     do {
       token = p.nextToken
       token match {
-        case OpenObj          => vals.push(JObject(Nil))
+        case OpenObj          => vals.push(IntermediateJObject(scala.collection.mutable.ListBuffer()))
         case FieldStart(name) => vals.push(JField(name, null))
         case StringVal(x)     => newValue(JString(x))
         case IntVal(x)        => newValue(JInt(x))
         case DoubleVal(x)     => newValue(JDouble(x))
         case BoolVal(x)       => newValue(JBool(x))
         case NullVal          => newValue(JNull)
         case CloseObj         => closeBlock(vals.popAny)
-        case OpenArr          => vals.push(JArray(Nil))
-        case CloseArr         => closeBlock(vals.pop(classOf[JArray]))
+        case OpenArr          => vals.push(IntermediateJArray(scala.collection.mutable.ListBuffer()))
+        case CloseArr         => closeBlock(vals.popAny)
         case End              =>
       }
     } while (token != End)
 
     root getOrElse JNothing
   }
 
-  private val EOF = (-1).asInstanceOf[Char]
+  private[this] final val EOF: Char = (-1).asInstanceOf[Char]
 
   private class ValStack(parser: Parser) {
-    import java.util.LinkedList
-    private[this] val stack = new LinkedList[Any]()
+    import java.util.ArrayDeque
+    private[this] val stack = new ArrayDeque[Any](32)
 
     def popAny = stack.poll
     def pop[A](expectedType: Class[A]) = convert(stack.poll, expectedType)
     def push(v: Any) = stack.addFirst(v)
     def peekAny = stack.peek
     def peek[A](expectedType: Class[A]) = convert(stack.peek, expectedType)
-    def replace[A](newTop: Any) = stack.set(0, newTop)
+    def replace[A](newTop: Any) = {
+      stack.pop
+      stack.push(newTop)
+    }
```

Comment:
If you mean the switch from `set(0, ...)` to `pop`/`push`, that's not what happened here. We switched to an [`ArrayDeque`](https://docs.oracle.com/javase/8/docs/api/java/util/ArrayDeque.html) instead of a `LinkedList`, and `ArrayDeque` doesn't support `set` operations, so we had to use `pop`/`push` instead.